docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,121 @@
+The applications are written in TypeScript but also have Wing files for infrastructure/deployment.
+
+To add a new application, the platform team needs to ensure that the application has the necessary permissions to interact with the shared resources. To do this, a kubernetes namespace and service account is created:
+
+```wing
+let namespace = cluster.createNamespace("app123");
+let serviceAccount = cluster.createServiceAccount(
+  namespace: namespace,
+  name: "app123-svc",
+);
+
+serviceAccount.grantRead(sharedBucket);
+```
+
+Applications using that namespace and service account will now have access to the shared bucket.
+
+To export this data to downstream applications, `cloud.Outputs` can be used:
+
+```wing
+let outputs = new cloud.Outputs();
+
+// Can reference preflight data (tokens will be resolved)
+outputs.add("namespace", namespace.name);
+outputs.add("serviceAccount", serviceAccount.name);
+
+// and inflight clients (includes the necessary information to interact with the resource for the given operations)
+// (e.g. a bucket name, a URL, etc.)
+outputs.lift("sharedBucket", sharedBucket, ["put"]);
+
+// Create a Wing library
+outputs.export(
+  name: "@acme/core-infra",
+);
+```
+
+This will create a centralized registry that can be imported elsewhere.
+During deployment, this registry will be updated with the relevant information.
+Additionally, it exports a wing library with all the necessary types and runtime code to look up that information.
+This library is published to an internal registry for other teams.
+
+After the central infrastructure is deployed, the application team can import the information into their own wing application:
+
+```wing
+bring "@acme/core-infra" as coreInfra;
+
+let workload = new Workload(
+  namespace: coreInfra.namespace,
+  serviceAccount: coreInfra.serviceAccount,
+  image: join(@dirname, "app"),
+);
+```
+
+Now the application will have the necessary permissions to interact with the shared bucket.
+
+However, we need to create a client for TypeScript to actually make use of these permissions.
+The environment where the client will run needs to know where to resolve the bucket.
+To do this, the client can attach to any inflight host. This takes care of preparing it for the client to run.
+Luckily, the workload exposes this host for us:
+
+```wing
+let client = new TypeScriptClient(
+  outdir: join(@dirname, ".wing"),
+  host: workload.host,
+);
+
+// No need to specify the operations, we want all the ones available from this limited resource
+client.lift("sharedBucket", coreInfra.sharedBucket);
+```
+
+After compilation, we can finally use this client in the TypeScript application:
+
+```ts
+// src/index.ts
+import { sharedBucket } from "../.wing";
+
+await sharedBucket.put("key", "value");
+```
+
+To run this locally, `wing it` is used for both the root infrastructure and then the application.
+
+### Implementation Thoughts
+
+#### Skipping cloud.Outputs?
```

Comment:
I don't think we need Outputs but we need a way to define a client and somehow retrieve (after/during deployment) the blob of "binding context" this client needs in order to be initialized, and leave the job of propagating this to the user.

```js
let client = new TypeScriptClient(...);


new CfnOutput(
  name: "ClientContext", 
  value: client.bindContext
);

// or

new k8s.ConfigMap(
  name: "ClientContext", 
  value: client.bindContext
)
```

Reply:
> leave the job of propagating this to the user.

This is basically 90% of the work involved in this feature.
If both the process of binding the data from the source and reading the data from the client requires explicit actions by the user, then all we're doing is generating types and writing our existing JS clients to disk with invisible requirements.

> we need a way to define a client and somehow retrieve (after/during deployment)

My proposal here is how I think that could work. Do you think it wouldn't? I understand your original proposal, I just don't know how it could actually work without giving the user a bad experience. I especially don't know how this would work with local development if the core infrastructure is running in the simulator.

The client can only resolve the necessary information in one way. Does it resolve it from somewhere on disk? environment variable?
If you intend to use a ConfigMap, they it would make sense to be on disk. If you're consuming this through a terraform/cfn output, then you might what it to be through environment variables. But since the client can only work one way, this choice only adds friction.

Reply:
See above comments.

We need to define the low level protocol and I think that should be the scope of this RFC.

It makes sense to include some examples in an appendix on how this low level protocol can be used to deploy apps in various environments.

Maybe include an example for:
1. K8s
2. Lambda
3. EC2

It would be valuable to look at three different examples to make sure the low level mechanism holds water.

But the goal is not to focus on the DX of each one of these examples, but rather to define the low level protocol.

Makes sense?